src/base64.cr

```diff
@@ -244,65 +286,159 @@ module Base64
+  # Processes the given data and yields each byte.
+  # The method does not raise when reading an invalid character.
+  # Only use this if you are certain the input is valid base64!
+  private def from_base64!(data : Bytes, &block : UInt8 -> Nil)
+    size = data.size
+    bytes = data.to_unsafe
+    bytes_begin = bytes
+
+    # Get the position of the last valid base64 character (rstrip '\n', '\r' and '=')
+    while size > 0 && (b = bytes[size - 1]) && (b == NL || b == NR || b == PAD)
+      size -= 1
+    end
+
+    # Process combinations of four characters until there aren't any left
+    fin = bytes + size - 4
+    while true
+      break if bytes > fin
+
+      # Move the pointer by one byte until there is a valid base64 character
+      while bytes.value == NL || bytes.value == NR
+        bytes += 1
+      end
+      break if bytes > fin
+
+      decode_chunk(
+        raise_error: false,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1], bytes[2], bytes[3]}
+      ) do |byte|
+        yield byte
+      end
+      bytes += 4
+    end
+
+    # Move the pointer by one byte until there is a valid base64 character or the end of `bytes` was reached
+    while (bytes < fin + 4) && (bytes.value == NL || bytes.value == NR)
+      bytes += 1
     end
-    res
+
+    # If the amount of base64 characters is not divisable by 4, the remainder of the previous loop is handled here
+    unread_bytes = (fin - bytes) % 4
+    case unread_bytes
+    when 1
+      raise Base64::Error.new("Wrong size")
+    when 2
+      decode_chunk(
+        raise_error: false,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1]}
+      ) do |byte|
+        yield byte
+      end
+    when 3
+      decode_chunk(
+        raise_error: false,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1], bytes[2]}
+      ) do |byte|
+        yield byte
+      end
+    end
+  end
+
+  # This macro decodes the given chunk of (2-4) base64 characters.
+  # The first argument (raise_error) determines if the macro raises an error when an invalid character is processed.
+  # The second argument (chunk_pos) is only used for the resulting error message.
+  # The resulting bytes are then each yielded to the given block, using the variable `byte`.
+  private macro decode_chunk(raise_error, chunk_pos, bytes, &block)
+    %buffer = 0_u32
+    {% for byte, i in bytes %}
+      %decoded = DECODE_TABLE[ {{byte}} ]
+      %buffer = (%buffer << 6) + %decoded
+      {% if raise_error == true %}
+        raise Base64::Error.new("Unexpected byte 0x#{{{byte}}.to_s(16)} at #{{{chunk_pos}} + {{i}}}") if %decoded == 255_u8
+      {% end %}
+    {% end %}
+
+    # Each byte in the buffer is shifted to rightmost position of the buffer, then casted to a UInt8
+    {% for i in 2..(bytes.size) %}
+      byte = (%buffer >> {{ (4 - bytes.size) * 2 + (8 * (bytes.size - i)) }}).to_u8!
```

Comment:
This will generate a whole bunch of overflow checks for each character. At the same time, we know it is safe (as long as the macro is only called with sufficiently small (2-4) `bytes`. Is there any noticeable performance change if these are changed to use the &-prefixed operators instead? Potentially the compiler may be smart enough to optimize away the checks, but who knows?

And if so, would we want that change?

(the comment also goes for the buffer assignment above).

Reply:
Just tested it, it's actually a little slower according to the benchmark (but all within margin of error, probably no difference)